File: app/src/main/java/net/ginteam/carmen/provider/auth/AuthProvider.java
package net.ginteam.carmen.provider.auth;

import net.ginteam.carmen.manager.ApiManager;
import net.ginteam.carmen.manager.PreferencesManager;
import net.ginteam.carmen.model.auth.AuthModel;
import net.ginteam.carmen.model.auth.UserModel;
import net.ginteam.carmen.network.api.service.AuthService;
import net.ginteam.carmen.network.api.subscriber.ModelSubscriber;
import net.ginteam.carmen.provider.ModelCallback;

import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by Eugene on 12/28/16.
 */

public class AuthProvider {

    private static AuthProvider sInstance;

    private AuthService mAuthService;
    private PreferencesManager mPreferencesManager;
    private UserModel mCachedUser;

    private AuthProvider() {
        mAuthService = ApiManager.getInstance().getService(AuthService.class);
        mPreferencesManager = PreferencesManager.getInstance();
    }

    public static AuthProvider getInstance() {
        if (sInstance == null) {
            sInstance = new AuthProvider();
        }
        return sInstance;
    }

    public UserModel getCurrentCachedUser() {
        return mCachedUser;
    }

    public void userLogin(final String email, final String password, final ModelCallback<UserModel> completion) {
        mAuthService
                .login(email, password)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new ModelSubscriber<AuthModel>() {
                    @Override
                    public void onSuccess(AuthModel resultModel) {
                        mCachedUser = resultModel.getUser();
                        mPreferencesManager.setUserToken(resultModel.getToken());
                        completion.onSuccess(mCachedUser);
                    }

                    @Override
                    public void onFailure(String message) {
                        mCachedUser = null;
                        completion.onFailure(message);
                    }
                });
    }

    public void userRegister(String name, final String email, final String password, final ModelCallback<UserModel> completion) {
        mAuthService
                .register(name, email, password)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new ModelSubscriber<AuthModel>() {
                    @Override
                    public void onSuccess(AuthModel resultModel) {
                        mCachedUser = resultModel.getUser();
                        mPreferencesManager.setUserToken(resultModel.getToken());
                        completion.onSuccess(mCachedUser);
                    }

                    @Override
                    public void onFailure(String message) {
                        mCachedUser = null;
                        completion.onFailure(message);
                    }
                });
    }

    public void fetchCurrentUser(final ModelCallback<UserModel> completion) {
        mAuthService
                .getCurrentUser()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new ModelSubscriber<UserModel>() {
                    @Override
                    public void onSuccess(UserModel resultModel) {
                        mCachedUser = resultModel;
                        completion.onSuccess(mCachedUser);
                    }

                    @Override
                    public void onFailure(String message) {
                        completion.onFailure(message);
                    }
                });
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/provider/AuthProvider.kt
package net.ginteam.carmen.kotlin.provider

import net.ginteam.carmen.kotlin.api.service.AuthService
import net.ginteam.carmen.kotlin.asyncWithCache
import net.ginteam.carmen.kotlin.manager.SharedPreferencesManager
import net.ginteam.carmen.kotlin.model.AuthModel
import net.ginteam.carmen.kotlin.model.ResponseModel
import net.ginteam.carmen.kotlin.model.UserModel
import net.ginteam.carmen.utils.DeviceUtils
import rx.Observable

/**
 * Created by eugene_shcherbinock on 2/10/17.
 */

interface AuthProvider {

    var currentCachedUser: UserModel?

    fun userLogin(email: String, password: String): Observable <ResponseModel <AuthModel>>
    fun userRegister(name: String, email: String, password: String): Observable <ResponseModel <AuthModel>>
    fun fetchCurrentUser(): Observable <ResponseModel <UserModel>>

    /*
        Device
     */

    fun deviceRegister(deviceId: String, pushToken: String, deviceType: String): Observable <ResponseModel <String>>
    fun updateNotificationStatus(notificationId: String): Observable <ResponseModel <String>>

}

object AuthenticationProvider : AuthProvider {
    private val authService: AuthService = AuthService.create()

    override var currentCachedUser: UserModel? = null

    override fun userLogin(email: String, password: String): Observable <ResponseModel <AuthModel>>
            = authService
            .userLogin(email, password, DeviceUtils.getDeviceId())
            .asyncWithCache(false)
            .doOnNext {
                cacheUser(it.data.user)
                saveUserAccessToken(it.data.token)
            }

    override fun userRegister(name: String, email: String, password: String): Observable <ResponseModel <AuthModel>>
            = authService
            .userRegister(name, email, password, DeviceUtils.getDeviceId())
            .asyncWithCache(false)
            .doOnNext {
                cacheUser(it.data.user)
                saveUserAccessToken(it.data.token)
            }

    override fun fetchCurrentUser(): Observable <ResponseModel <UserModel>>
            = authService.getCurrentUser().asyncWithCache().doOnNext { cacheUser(it.data) }

    override fun deviceRegister(deviceId: String, pushToken: String, deviceType: String): Observable<ResponseModel<String>>
            = authService.deviceRegister(deviceId, pushToken, deviceType).asyncWithCache()

    override fun updateNotificationStatus(notificationId: String): Observable <ResponseModel <String>>
            = authService.updateNotificationStatus(notificationId).asyncWithCache()

    private fun cacheUser(user: UserModel) {
        currentCachedUser = user
    }

    private fun saveUserAccessToken(token: String) {
        SharedPreferencesManager.userAccessToken = token
    }
}